### ISSUE TITLE
End-to-end serverless workflow

### ISSUE THREAD

**Issue by danieldreier** [hashicorp/terraform-cdk#733]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
This is a draft! Let's talk before starting work on this.

The purpose of this project is to design a good end-to-end workflow for building and deploying serverless applications to popular cloud providers. This is simultaneously a documentation and design exercise. In the first iteration, the documentation should describe how to implement these workflows using CDK for Terraform as it exists today.

While writing that documentation, consider what else we could do to simplify the workflow and make it more idiomatic in each language and cloud provider. File enhancement requests, bugs, and link them back to this issue.

Use case: user wants to deploy a simple serverless application without having to stand up or operate other infrastructure.

#### Goals
* All resources are defined in a single codebase
* Examples available for AWS Lambda, Azure Functions, and (stretch goal) Google Cloud Functions
* Deploy functions, load balancers, security groups, storage buckets etc as needed - no manual setup outside of `cdktf deploy` should be needed.
* The serverless functions, as well as the CDK for Terraform code examples, should be implemented in TypeScript, Python, and Go

#### Stretch goals
As time permits, more advanced permutations would also be helpful, such as:
* deploying and configuring persistence (db, storage buckets) including setting up authentication to the persistence
* separately deploying a toy backend API and front-end as separate containers
* Provide a clear developer workflow for publishing to development/staging/production environments
* When the deploy completes, show (using an output?) the URL the web application is available at

#### Ideas for sample applications
In order to illustrate serverless use cases, we need some concrete use cases that are representative of common scenarios. Below, I've proposed some ideas. I'll leave it to whoever picks up this ticket to define the exact cases. Please pick one or two and add a more detailed definition (edit this issue as needed) prior to starting work.

* Simple REST endpoint that uses the cloud provider's key-value store for persistence
  * possibly accompanied by a simple statically hosted (cloud storage bucket) front-end app that pulls data from the endpoint, to show how both can be deployed as separate stacks
* Recurring (e.g. cron job) lambda that polls something and stores the result (probably in cloud provider's key-value store)
* Simple file processing pipeline where:
  * a file is POSTed to a first HTTP endpoint, which saves it to a cloud storage bucket
  * the upload event triggers another function, which compresses (gzip is fine) the file and deletes the original uncompressed file
  * a third HTTP endpoint shows a list of compressed files which are available for download
  * every day any compressed files over 24 hours old are deleted

Please feel free to suggest other toy use cases instead of these! This is intended as a prompt.

### Deliverables
The immediate practical output will be end-to-end documentation published in this GitHub. File any ideas you have for improving it as enhancement requests linked to this issue for ease of tracking.
